### PR TITLE
Add cache support (again).

### PR DESCRIPTION
**Pull request by pmclanahan**
This is another try at this, and should replace #32, which I don't seem to be
able to reopen now that I've force-pushed my branch.

* Make it more testable and add basic tests.
* Add tox for testing in 2.6 and 2.7 with Django 1.6 - 1.8.
* Have Travis-CI run the tox configs.

### CODE REVIEW

**Comment by pmclanahan** [comment]
I've updated things, improved Tox config, and added Travis (though I don't seem to be able to enable the repo in Travis at the moment). Let's see if we can land this thing.

@jezdez @fwenzel @jsocol: you were the ones with strong opinions on #32. Any of you have time for one more r?

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+[testenv]
+usedevelop = true
+pip_pre = true
+commands = nosetests
+deps =
+    1.6: Django>=1.6,<1.7
+    1.7: Django>=1.7,<1.8
+    1.8: Django>=1.8,<1.9
+    mock==1.0.1
+    nose==1.3.3
```
Comment (jezdez) [inline tox.ini]: I think nose has a newer version out.

File: requirements.txt [hunk 2]
```diff
@@ -1,2 +1,2 @@
 # Requirements for building tests
-nose
+tox==1.7.2
```
Comment (jezdez) [inline requirements.txt]: tox has a newer version

**Comment by jezdez** [comment]
rwc+

File: tox.ini (hunk 1 above)
Comment (pmclanahan) [inline tox.ini]: Probably okay to just leave the version off for these?

**Comment by pmclanahan** [comment]
Thanks jezdez. Updated versions.

**Comment by pmclanahan** [comment]
@jgmize care to r?

**Comment by pmclanahan** [comment]
I've now done some testing with bedrock and updated the cache mechanism to also cache a file miss, and added a test for this.

**Comment by pmclanahan** [comment]
One thing that I've realized is that we kindof have the same issue we had that was fixed in #37 because all instances and instances of subclasses will use the same cache. I think this is more of a feature than a bug, but I want more opinions. It will likely require some reworking of tests if they're relying on mocking of product details (e.g. bedrock will need some test reworking).

**Comment by jgmize** [comment]
> One thing that I've realized is that we kindof have the same issue we had that was fixed in #37 because all instances and instances of subclasses will use the same cache. I think this is more of a feature than a bug, but I want more opinions. 

As far as I can tell, the data in the cache isn't mutated by the various subclasses we have now, so a shared cache is *probably* fine, and for hypothetical use case of a subclass that needs to mutate the cache for some reason could override the `_cache_key` class attr or the `_get_cache_key` method to use a separate cache for that, so I don't think that should necessarily block this PR. I think the real issues stem from using subclasses for this purpose in bedrock in the first place, and I'd be happy to refactor their usage out in favor of a more functional approach later.

**Comment by jgmize** [comment]
@pmclanahan I believe you've addressed everything from @jezdez's original r+wc, and this lgtm as well. I'll leave it up to you to merge, in case you wanted to do any squashing of the commits or any other cleanup.

**Comment by pmclanahan** [comment]
Thanks @jgmize.

**Pull request merged**